### PR TITLE
Use ff crate for traits and impls

### PR DESCRIPTION
**Pull request by str4d**
Depends on https://github.com/ebfull/ff/pull/1 and https://github.com/ebfull/ff/pull/5

### CODE REVIEW

**Comment by ebfull** [comment]
ff has been updated to `0.3.0`.

File: Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,10 @@ rand = "0.4"
 byteorder = "1"
 clippy = { version = "0.0.200", optional = true }
 
+[dependencies.ff]
+git = "https://github.com/str4d/ff"
+rev = "91a8dc8284a09c456b36da44ca7de36f8b50550d"
+
 [features]
```
Comment (ebfull) [inline Cargo.toml]: Let's change `u128-support` feature in here to `u128-support = ["ff/u128-support"]` until we've removed `u128-support` from `ff`.

**Comment by ebfull** [comment]
Add yourself to `Cargo.toml`::`authors` because I have a feeling this isn't the last PR you're submitting!

**Comment by str4d** [comment]
Comments addressed

File: src/bls12_381/fq.rs [hunk 2]
```diff
@@ -1,69 +1,5 @@
...
-// 2^s root of unity computed by GENERATOR^t
-const ROOT_OF_UNITY: FqRepr = FqRepr([
-    0x43f5fffffffcaaae,
-    0x32b7fff2ed47fffd,
-    0x7e83a49a2e99d69,
-    0xeca8f3318332bb7a,
-    0xef148d1ea0f4c069,
-    0x40ab3263eff0206,
-]);
+use {Field, PrimeField, PrimeFieldDecodingError, PrimeFieldRepr};
```
Comment (ebfull) [inline src/bls12_381/fq.rs]: `use ff::{`

**Comment by ebfull** [comment]
@bmerge try

**Comment by bmerge** [comment]
:hourglass: Trying commit c49590bab72161706343e584d839a55e8913f799 with merge c123b19ee4840f2c3274f3634fc4ad81a00f4b0c...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux64-try](http://ci.bellman.rs:8010/builders/pairing-linux64-try/builds/13)

**Comment by ebfull** [comment]
Updated CI to `nightly-2018-06-28`.

@bmerge retry

**Comment by bmerge** [comment]
:hourglass: Trying commit c49590bab72161706343e584d839a55e8913f799 with merge 2e1885d0d9d1b51d446f3663b60bf0f37aa85f25...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux64-try](http://ci.bellman.rs:8010/builders/pairing-linux64-try/builds/14)

**Comment by ebfull** [comment]
@bmerge try

**Comment by bmerge** [comment]
:hourglass: Trying commit 095483638496d896435b99c70e8ac0025411a96c with merge 69a88a6902f6e0916c1cf229bf3118507c81745a...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux64-try](http://ci.bellman.rs:8010/builders/pairing-linux64-try/builds/15)

**Comment by ebfull** [comment]
@bmerge try

**Comment by bmerge** [comment]
:hourglass: Trying commit 5e35ba44bed94ccc14949f607652d83ec4f59e1f with merge 0a53b45f326d524e9c94d1296b1333498460631d...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux32-try](http://ci.bellman.rs:8010/builders/pairing-linux32-try/builds/16)

**Comment by ebfull** [comment]
@bmerge retry

**Comment by bmerge** [comment]
:hourglass: Trying commit 5e35ba44bed94ccc14949f607652d83ec4f59e1f with merge c5b78506edcbf5f9074aff6bd9134eac6571c8c6...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-windows32msvc-try](http://ci.bellman.rs:8010/builders/pairing-windows32msvc-try/builds/17)

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux64-try](http://ci.bellman.rs:8010/builders/pairing-linux64-try/builds/18)

**Comment by ebfull** [comment]
@bmerge try

**Comment by bmerge** [comment]
:hourglass: Trying commit 9a9150d869eecea29a7f040c58a7caa900bc67c7 with merge 8b79997d70b7178f21590faee062d0e181b40def...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux32-try](http://ci.bellman.rs:8010/builders/pairing-linux32-try/builds/19)

**Comment by ebfull** [comment]
@bmerge retry

**Comment by bmerge** [comment]
:hourglass: Trying commit 9a9150d869eecea29a7f040c58a7caa900bc67c7 with merge ffb636e488fdda62dbeb1b3ce8c24b8c8cc75724...

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux64-try](http://ci.bellman.rs:8010/builders/pairing-linux64-try/builds/20)

**Comment by bmerge** [comment]
:broken_heart: Test failed - [pairing-linux64-try](http://ci.bellman.rs:8010/builders/pairing-linux64-try/builds/21)

**Comment by ebfull** [comment]
@bmerge try

**Comment by bmerge** [comment]
:hourglass: Trying commit b140f61b027f9a16325fa751bb92bc015f85422e with merge 13799c746b0cdab051102c1e399aa89defbe67d3...

**Comment by bmerge** [comment]
:sunny: Test successful - [pairing-linux32-try](http://ci.bellman.rs:8010/builders/pairing-linux32-try/builds/22), [pairing-linux64-try](http://ci.bellman.rs:8010/builders/pairing-linux64-try/builds/22), [pairing-windows32msvc-try](http://ci.bellman.rs:8010/builders/pairing-windows32msvc-try/builds/22), [pairing-windows64msvc-try](http://ci.bellman.rs:8010/builders/pairing-windows64msvc-try/builds/22)
State: approved= try=True

**Comment by str4d** [comment]
Squashed a couple of the later commits together.

**Comment by ebfull** [comment]
@bmerge try

**Comment by bmerge** [comment]
:hourglass: Trying commit fa8103764a07bd273927447d434de18aace252d3 with merge df50d7d8a17c42e6ef953158083418f2a776c988...

**Comment by bmerge** [comment]
:sunny: Test successful - [pairing-linux32-try](http://ci.bellman.rs:8010/builders/pairing-linux32-try/builds/23), [pairing-linux64-try](http://ci.bellman.rs:8010/builders/pairing-linux64-try/builds/23), [pairing-windows32msvc-try](http://ci.bellman.rs:8010/builders/pairing-windows32msvc-try/builds/23), [pairing-windows64msvc-try](http://ci.bellman.rs:8010/builders/pairing-windows64msvc-try/builds/23)
State: approved= try=True

**Comment by bmerge** [comment]
:sunny: Test successful - [pairing-linux32-auto](http://ci.bellman.rs:8010/builders/pairing-linux32-auto/builds/57), [pairing-linux64-auto](http://ci.bellman.rs:8010/builders/pairing-linux64-auto/builds/62), [pairing-windows32msvc-auto](http://ci.bellman.rs:8010/builders/pairing-windows32msvc-auto/builds/57), [pairing-windows64msvc-auto](http://ci.bellman.rs:8010/builders/pairing-windows64msvc-auto/builds/57)
Approved by: ebfull
Pushing 183a64b08e9dc7067f78624ec161371f1829623e to master...

**Pull request merged**